ludwig/schema/preprocessing.py

```diff
@@ -0,0 +1,699 @@
+        MISSING_VALUE_STRATEGY_OPTIONS,
+        default="fill_with_const",
+        allow_none=False,
+        description="What strategy to follow when there's a missing value in a text column",
+    )
+
+    fill_value: Optional[str] = schema_utils.String(
+        default="<UNK>",
+        allow_none=False,
+        description="The value to replace missing values with in case the missing_value_strategy is fill_with_const",
+    )
+
+
+@dataclass
+@register_preprocessor(NUMBER)
+class NumberPreprocessingConfig(schema_utils.BaseMarshmallowConfig):
+    """NumberPreprocessingConfig is a dataclass that configures the parameters used for a number input feature."""
+
+    missing_value_strategy: Optional[str] = schema_utils.StringOptions(
+        MISSING_VALUE_STRATEGY_OPTIONS,
+        default="fill_with_const",
+        allow_none=False,
+        description="What strategy to follow when there's a missing value in a number column",
+    )
+
+    fill_value: Optional[float] = schema_utils.NonNegativeFloat(
+        default=0.0,
+        allow_none=False,
+        description="The value to replace missing values with in case the missing_value_strategy is fill_with_const",
+    )
+
+    normalization: Optional[str] = schema_utils.StringOptions(
+        ["zscore", "minmax", "log1p"],
+        default=None,
+        allow_none=True,
+        description="Normalization strategy to use for this number feature.",
+    )
+
+
+@dataclass
+@register_preprocessor(BINARY)
+class BinaryPreprocessingConfig(schema_utils.BaseMarshmallowConfig):
+    """BinaryPreprocessingConfig is a dataclass that configures the parameters used for a binary input feature."""
+
+    missing_value_strategy: Optional[str] = schema_utils.StringOptions(
+        MISSING_VALUE_STRATEGY_OPTIONS + ["fill_with_false"],
+        default="fill_with_false",
+        allow_none=False,
+        description="What strategy to follow when there's a missing value in a binary column",
+    )
+
+    fill_value: Union[int, float] = schema_utils.NumericOrStringOptionsField(
+        strings_utils.all_bool_strs(),
+        default=None,
+        default_numeric=None,
+        default_option=None,
+        allow_none=False,
+        min=0,
+        max=1,
+        description="The value to replace missing values with in case the missing_value_strategy is fill_with_const",
+    )
+
+    fallback_true_label: Optional[str] = schema_utils.String(
+        default=None,
+        allow_none=True,
+        description="The label to interpret as 1 (True) when the binary feature doesn't have a "
+        "conventional boolean value",
+    )
+
+
+@dataclass
+@register_preprocessor(CATEGORY)
+class CategoryPreprocessingConfig(schema_utils.BaseMarshmallowConfig):
+    """CategoryPreprocessingConfig is a dataclass that configures the parameters used for a category input
+    feature."""
+
+    missing_value_strategy: Optional[str] = schema_utils.StringOptions(
```

Comment:
The rationale here is that this is a required field in the config, but it can't be `None`. If the user doesn't touch it, then it populates with the default, which is why it's typed as `Optional[str]` but if they do decide to touch it, they can't set the value to None, otherwise we'll get an "Invalid missing_value_strategy error". So the solution was to set it to `None` to avoid the default.

Reply:
Might be misunderstanding you, but the default choice is `fill_with_const`, right? Imagining I delete `Optional` for a moment and then the functionality is: if a user types nothing, `fill_with_const` is what gets chosen; at the same time, setting `allow_none` to `False` prevents the user from typing `missing_value_strategy: null` in the config. So `missing_value_strategy` can never be `None`. That seems like what we want, right?